### PR TITLE
Color speed in hud based on increase or decrease

### PR DESCRIPTION
**Pull request by ChillerDragon**
https://github.com/user-attachments/assets/acb54e53-6bd7-47b7-88d1-e98b7b64f125

<details>
<summary>Outdated colors preview</summary>

https://github.com/user-attachments/assets/5faf0ebf-4799-4b79-81ae-7e81e45b13e6

</details>

Sometimes the Y velocity is marked green when walking on the floor this is caused by the server bug described in https://github.com/ddnet/ddnet/pull/8743

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by l-ouis** [comment]
it would be nicer if the red color was closer to white (lighter) since it's pretty hard to see

File: src/game/client/components/hud.cpp [hunk 1]
```diff
@@ -1311,7 +1312,7 @@ void CHud::RenderMovementInformation(const int ClientId)
 	// To make the player position relative to blocks we need to divide by the block size
 	const vec2 Pos = mix(vec2(pPrevChar->m_X, pPrevChar->m_Y), vec2(pCurChar->m_X, pCurChar->m_Y), IntraTick) / 32.0f;
 
-	const vec2 Vel = mix(vec2(pPrevChar->m_VelX, pPrevChar->m_VelY), vec2(pCurChar->m_VelX, pCurChar->m_VelY), IntraTick);
+	const ivec2 Vel = mix(ivec2(pPrevChar->m_VelX, pPrevChar->m_VelY), ivec2(pCurChar->m_VelX, pCurChar->m_VelY), IntraTick);
```
Comment (heinrich5991) [inline src/game/client/components/hud.cpp:1315]: Why was this changed to integers?

File: src/game/client/components/hud.cpp [hunk 2]
```diff
@@ -1364,15 +1365,36 @@ void CHud::RenderMovementInformation(const int ClientId)
...
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 0, 0, 1);
+			}
+			if(AbsVel < 2)
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 1, 1, 1);
+			}
+			ForceUpdate = m_aPlayerSpeed[i] != AbsVel;
+			m_aPlayerSpeed[i] = AbsVel;
+			TextRender()->TextColor(m_aPlayerSpeedColor[i]);
```
Comment (heinrich5991) [inline src/game/client/components/hud.cpp:1389]: This should probably go into the the `UpdateMovementInformationTextContainer`.

File: src/game/client/components/hud.cpp [hunk 3]
```diff
@@ -1259,12 +1260,12 @@ inline float CHud::GetMovementInformationBoxHeight()
 	return BoxHeight;
 }
 
-void CHud::UpdateMovementInformationTextContainer(STextContainerIndex &TextContainer, float FontSize, float Value, char *pPrevValue, size_t Size)
+void CHud::UpdateMovementInformationTextContainer(STextContainerIndex &TextContainer, float FontSize, float Value, char *pPrevValue, size_t Size, bool ForceUpdate)
```
Comment (heinrich5991) [inline src/game/client/components/hud.cpp:1263]:
```suggestion
void CHud::UpdateMovementInformationTextContainer(STextContainerIndex &TextContainer, float FontSize, float Value, char *pPrevValue, size_t Size)
```
It's unclear what `ForceUpdate` is supposed to be, and it's probably an antipattern. The function should recognize that something changed and update it based on that.

File: src/game/client/components/hud.cpp [hunk 4]
```diff
@@ -1364,15 +1365,36 @@ void CHud::RenderMovementInformation(const int ClientId)
...
+			if(AbsVel < m_aPlayerSpeed[i])
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 0, 0, 1);
+			}
+			if(AbsVel < 2)
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 1, 1, 1);
+			}
+			ForceUpdate = m_aPlayerSpeed[i] != AbsVel;
+			m_aPlayerSpeed[i] = AbsVel;
```
Comment (heinrich5991) [inline src/game/client/components/hud.cpp:1388]: Having this logic in a `Render` function seems brittle. This means the values depend on how often `Render` is called. Can you change it so that the `Render` function isn't stateful?

File: src/game/client/components/hud.cpp (hunk 1 above)
Comment (ChillerDragon) [reply src/game/client/components/hud.cpp:1315]: `pPrevChar->m_VelX` is an integer. And I don't trust float comparisons. If that is a problem I guess I could also cast it to an int in the comparison.

File: src/game/client/components/hud.cpp [hunk 5]
```diff
@@ -1364,15 +1365,36 @@ void CHud::RenderMovementInformation(const int ClientId)
...
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(0, 1, 0, 1);
+			}
+			if(AbsVel < m_aPlayerSpeed[i])
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 0, 0, 1);
+			}
+			if(AbsVel < 2)
+			{
+				m_aPlayerSpeedColor[i] = ColorRGBA(1, 1, 1, 1);
```
[the remark this answers is not in the document]
Comment (ChillerDragon) [reply src/game/client/components/hud.cpp:1385]: The use case is for racers to see if a certain move helped their speed or not. [corneum](https://ddnet.org/players/corneum/) requested that feature once.

**Comment by ChillerDragon** [comment]
Softer red:

https://github.com/user-attachments/assets/acb54e53-6bd7-47b7-88d1-e98b7b64f125

**Review by def- — APPROVED** [review]

**Comment by def-** [comment]
Why does it stay green when you move constantly in the same direction? Shouldn't that be white?

**Comment by def-** [comment]
Also, the colors could be even closer to white.

**Comment by ChillerDragon** [comment]
> Why does it stay green when you move constantly in the same direction? Shouldn't that be white?

It shows if the last speed change was increasing or decreasing not if it is increasing or decreasing right now. Otherwise it is changing too fast to look at. So you can see if for example the last hook you made helped your speed or not.

**Comment by def-** [comment]
(Sounds a bit like you would want a graph of your acceleration instead)

**Comment by ChillerDragon** [comment]
> (Sounds a bit like you would want a graph of your acceleration instead)

Yea sounds good. The problem here is that I personally do not care about this feature at all. I am not a racer. Its just that a player requested it and it was on my todo to add it to ddnet. Maybe we can get a review from a player who would actually use it.
